vault/resource_managed_keys.go

```diff
@@ -0,0 +1,646 @@
+	"github.com/hashicorp/terraform-provider-vault/internal/provider"
+)
+
+const (
+	kmsTypePKCS  = "pkcs11"
+	kmsTypeAWS   = "awskms"
+	kmsTypeAzure = "azurekeyvault"
+)
+
+func managedKeysResource() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: createUpdateManagedKeys,
+		DeleteContext: deleteManagedKeys,
+		ReadContext:   readManagedKeys,
+		UpdateContext: createUpdateManagedKeys,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			consts.FieldPKCS: {
+				Type:        schema.TypeSet,
+				Optional:    true,
+				Description: "Configuration block for PKCS Managed Keys",
+				Elem: &schema.Resource{
+					Schema: managedKeysPKCSConfigSchema(),
+				},
+				Set: hashManagedKeys,
+			},
+			consts.FieldAWS: {
+				Type:        schema.TypeSet,
+				Optional:    true,
+				Description: "Configuration block for AWS Managed Keys",
+				Elem: &schema.Resource{
+					Schema: managedKeysAWSConfigSchema(),
+				},
+				Set: hashManagedKeys,
+			},
+			consts.FieldAzure: {
+				Type:        schema.TypeSet,
+				Optional:    true,
+				Description: "Configuration block for Azure Managed Keys",
+				Elem: &schema.Resource{
+					Schema: managedKeysAzureConfigSchema(),
+				},
+				Set: hashManagedKeys,
+			},
+		},
+	}
+}
+
+func hashManagedKeys(v interface{}) int {
+	var result int
+	if m, ok := v.(map[string]interface{}); ok {
+		if v, ok := m[consts.FieldName]; ok {
+			result = helper.HashCodeString(v.(string))
+		}
+	}
+
+	return result
+}
+
+func getCommonManagedKeysSchema() schemaMap {
+	return schemaMap{
+		consts.FieldAllowGenerateKey: {
+			Type:     schema.TypeBool,
+			Optional: true,
+			Computed: true,
+			Description: "If no existing key can be found in the referenced " +
+				"backend, instructs Vault to generate a key within the backend",
+		},
+
+		consts.FieldAllowReplaceKey: {
+			Type:     schema.TypeBool,
+			Optional: true,
+			Computed: true,
+			Description: "Controls the ability for Vault to replace through " +
+				"generation or importing a key into the configured backend even " +
+				"if a key is present, if set to false those operations are forbidden " +
+				"if a key exists.",
+		},
+
+		consts.FieldAllowStoreKey: {
+			Type:     schema.TypeBool,
+			Optional: true,
+			Computed: true,
+			Description: "Controls the ability for Vault to import a key to the " +
+				"configured backend, if 'false', those operations will be forbidden",
+		},
+
+		consts.FieldAnyMount: {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Computed:    true,
+			Description: "Allow usage from any mount point within the namespace if 'true'",
+		},
+
+		consts.FieldUUID: {
+			Type:        schema.TypeString,
+			Computed:    true,
+			Description: "ID of the managed key read from Vault",
+		},
+	}
+}
+
+func setCommonManagedKeysSchema(s schemaMap) schemaMap {
+	for k, v := range getCommonManagedKeysSchema() {
```

Comment:
I think you probably want  https://github.com/hashicorp/terraform-provider-vault/blob/aaa7b1a5ba8caeba1ed1c8fd05a49ef0f695ba26/vault/provider.go#L822, which will panic on duplicate fields. Or add your own check for `key` duplication.

Reply:
Ah nice catch! I added a key duplication check in my function; I didn't want to change all the schema function signatures to match the `mustAddSchema` signature, but the duplication was a quick fix :)